### ISSUE TITLE
deserialization fails if environment contains multiple attributes or namespace prefixes

### ISSUE THREAD

**Issue by impact-merlinmarek** [aas-core-works/aas-core3.0-golang#25]
Some of the aas files we encounter use multiple attributes in the environment tag. This is a minimal example:

```xml
<aas:environment xmlns:aas="https://admin-shell.io/aas/3/0" xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance" xsi:schemaLocation="https://admin-shell.io/aas/3/0 AAS.xsd">
</aas:environment>
```

This validates against the official xml schema and can be useful besides specifying the schema, e.g. when using elements from additional namespaces. When parsing the xml with the golang sdk, it fails with:

```
Expected no attributes except 'xmlns' in the start element, but got 3 in the start element environment
```

If we remove all attributes except the aas namespace, parsing still fails because the xmlns is not raw but uses a prefix:

```xml
<aas:environment xmlns:aas="https://admin-shell.io/aas/3/0">
</aas:environment>
```
=>
```
Expected no attributes except 'xmlns' in the start element, but got 1 in the start element environment
```

Could you add support for these cases in the sdk? Thanks in advance.

**Comment by mristin** [comment]
Hi @impact-merlinmarek ,
> If we remove all attributes except the aas namespace, parsing still fails because the xmlns is not raw but uses a prefix:
> ```
> <aas:environment xmlns:aas="https://admin-shell.io/aas/3/0">
> </aas:environment>
> ```

I think your XML is invalid, even though it passes the schema engine. Mind that many schema engines are finicky and often allow invalid XMLs.

As far as I know, you need to define the the XML namespace at the root element, and then it applies to children elements. What you are doing here is defining an XML *alias*, which is actually futile.

You would use the aliasing if you are embedding your XML in a larger document:
```
<someLargerDocument xmlns:aas="https://admin-shell.io/aas/3/0">
<aas:environment>...</aas:environment>
</someLargerDocument>
```

> Could you add support for these cases in the sdk?

I'm a bit lost which attributes should be allowed and expected in the root element. Introducing aliases is certainly valid (`xmlns:Xxx`). Should `xsi:Xxx` also be allowed?

I found the following two related StackOverflow questions:
* https://stackoverflow.com/questions/34202967/xmlns-xmlnsxsi-xsischemalocation-and-targetnamespace
* https://stackoverflow.com/questions/34202967/xmlns-xmlnsxsi-xsischemalocation-and-targetnamespace

Can you please also look on Internet -- as soon as we nail down which attributes are commonly expected, I'd fix the SDK to simply ignore them in the elements. 

As far as I understand, [xml.Decoder](https://pkg.go.dev/encoding/xml#Decoder) takes care of the namespaces, so the attributes will be handled by the XMLDecoder:
https://pkg.go.dev/encoding/xml#Name

> type [Name](https://cs.opensource.google/go/go/+/go1.22.3:src/encoding/xml/xml.go;l=40) [¶](https://pkg.go.dev/encoding/xml#Name)
>
> ```
> type Name struct {
>	Space, Local [string](https://pkg.go.dev/builtin#string)
> }
> ```
>
> A Name represents an XML name (Local) annotated with a name space identifier (Space). In tokens returned by [Decoder.Token](https://pkg.go.dev/encoding/xml#Decoder.Token), the Space identifier is given as a canonical URL, not the short prefix used in the document being parsed.

**Comment by impact-merlinmarek** [comment]
I will do some research. Could you post the link to the second stackoverflow question again please? Both links in your post are identical.

**Comment by mristin** [comment]
🤦‍♂️ I think it was this one:
https://stackoverflow.com/questions/41035128/what-is-the-difference-between-xsd-and-xsi

**Comment by impact-merlinmarek** [comment]
> I think your XML is invalid, even though it passes the schema engine. Mind that many schema engines are finicky and often allow invalid XMLs.

Could you elaborate on why you think this xml is invalid? Do you know of a strict xml validator for xml files?

> As far as I know, you need to define the the XML namespace at the root element, and then it applies to children elements. What you are doing here is defining an XML _alias_, which is actually futile.

The aas file this is pulled from uses the `aas` prefix for all elements explicitly which semantically has the same effect as a default namespace and element declarations without a namespace prefix.

> I'm a bit lost which attributes should be allowed and expected in the root element. Introducing aliases is certainly valid (`xmlns:Xxx`). Should `xsi:Xxx` also be allowed?

The XML specification does not restrict the usage of a colon in attribute names but discourages usage except for namespace purposes (https://www.w3.org/TR/xml/ 2.3):

> Note:
> 
> The Namespaces in XML Recommendation [XML Names] assigns a meaning to names containing colon characters. Therefore, authors should not use the colon in XML names except for namespace purposes, but XML processors must accept the colon as a name character.

Following this, I think all attribute names that conform to the XML spec should be allowed.

> Can you please also look on Internet -- as soon as we nail down which attributes are commonly expected, I'd fix the SDK to simply ignore them in the elements.

These are all the different `environment` start elements that I have observed so far:

```
<aas:environment xmlns:aas="https://admin-shell.io/aas/3/0" xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance" xsi:schemaLocation="https://admin-shell.io/aas/3/0 AAS.xsd">

<aas:environment xmlns:aas="https://admin-shell.io/aas/3/0">

<environment xmlns="https://admin-shell.io/aas/3/0">
```

**Comment by mristin** [comment]
>> I think your XML is invalid, even though it passes the schema engine. Mind that many schema engines are finicky and often allow invalid XMLs.
>
> Could you elaborate on why you think this xml is invalid? 

I'm by no means XML specialist -- in my understanding, you only define the namespace alias -- but it could be that it is also defined backwards, so that you can use it in the element where you define it.

Which is also in line with what you wrote below:

> The aas file this is pulled from uses the aas prefix for all elements explicitly which semantically has the same effect as a default namespace and element declarations without a namespace prefix.

> Do you know of a strict xml validator for xml files?

Unfortunately, I do not. I only observed that some of the engines I used as libraries in C#, Python and Java as well as some on-line verificators had different behaviors.

> Following this, I think all attribute names that conform to the XML spec should be allowed.

The AAS specs allow *no* XML attributes. We need therefore to define the set of pre-defined XML attributes which should be ignored. In my understanding, that would be attributes `xmlns:Xxx` and `xsi:Xxx` -- but are there more such prefixes which should be ignored?

In other words, are those prefixes of XML attribute names (`xmlns:` and `xsi:`) specified anywhere?

**Comment by impact-merlinmarek** [comment]
> I'm by no means XML specialist -- in my understanding, you only define the namespace alias -- but it could be that it is also defined backwards, so that you can use it in the element where you define it.

The xml namespace specification allows to use the namespace alias in the same element that declares it (https://www.w3.org/TR/xml-names/ 5):

> An example of a qualified name serving as an element name:
> ```xml
> 
> <edi:price xmlns:edi='http://ecommerce.example.org/schema' units='Euro'>32.18</edi:price>
> ```
>
> Namespace constraint: Prefix Declared
> 
> The namespace prefix, unless it is xml or xmlns, MUST have been declared in a namespace declaration attribute in either the start-tag of the element where the prefix is used or in an ancestor element (i.e., an element in whose content the prefixed markup occurs).

> The AAS specs allow _no_ XML attributes.

Could you point me to the relevant specification?

> We need therefore to define the set of pre-defined XML attributes which should be ignored. In my understanding, that would be attributes `xmlns:Xxx` and `xsi:Xxx` -- but are there more such prefixes which should be ignored?
> 
> In other words, are those prefixes of XML attribute names (`xmlns:` and `xsi:`) specified anywhere?

`xmlns` and `xmlns:` are declared in https://www.w3.org/TR/xml-names/ 3.

I think `xsi:` is explained in https://www.w3.org/TR/xmlschema11-1/ (*I have not read this one*).

**Comment by impact-merlinmarek** [comment]
I think namespacing and automatic schema verification are reasonable use cases, so I would be in favor of allowing those.

I am obviously biased here but I think that the sdk would generally be more useful as a software building block if most errors are pushed to the verification phase. As an example, additional attributes generally do not pose a problem when trying to extract information from an aas xml. 

Currently most aasx files that I have seen online need a preprocessing phase before they can be parsed although they seem good enough (while violating the spec) to extract useful information.

Of course it would be nice if all vendors would respect the spec. I think one step in this direction would be to adjust the official xsd to enforce the same restrictions as the sdk (_if possible, I have little experience with xml schemas_).

**Comment by mristin** [comment]
@impact-merlinmarek I finally decided to go for a compromise: `xmlns` attributes are ignored as their semantics is clearly specified and unambiguous.

For cases where your XML contains other attributes such as `xsi:Xxx`, you have to remove the attributes yourself. You have an example snippet here: https://github.com/aas-core-works/aas-core3.0-golang?tab=readme-ov-file#no-attributes-expected. In particular, I wanted to avoid a piling-up of prefixes to ignore, and wanted to have a solution for a larger family of cases.

I hope this works out in the practice.

**Comment by mristin** [comment]
P.S. If you happen to find where `xsi` is officially defined for XML documents, I'll gladly add it to the ignored attributes. I suppose the bar for inclusion should be that official W3C documentation mentions it somewhere.

**Comment by impact-merlinmarek** [comment]
Thank you for implementing this so fast. Since `xsi` is not relevant for parsing I think we can just drop it while parsing using a custom `xml.TokenReader`.
